<a href="https://colab.research.google.com/github/mib-a/AI-ML-Learning/blob/main/mib-a/AI-ML-Learning/Month-03-Deep-Learning/Module-11/Breast_Cancer_Python_Training_Pipeline.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 11.2 Without inheriting nn module (Basic Way)

## Data Loading and Basic Preprocessing

## Import Libraries

In [581]:
import numpy as np
import pandas as pd
import torch
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.preprocessing import LabelEncoder

In [582]:
df = pd.read_csv('https://raw.githubusercontent.com/gscdit/Breast-Cancer-Detection/refs/heads/master/data.csv')
df.head()

,id,diagnosis,radius_mean,texture_mean,perimeter_mean,area_mean,smoothness_mean,compactness_mean,concavity_mean,concave points_mean,...,texture_worst,perimeter_worst,area_worst,smoothness_worst,compactness_worst,concavity_worst,concave points_worst,symmetry_worst,fractal_dimension_worst,Unnamed: 32
0,842302,M,17.99,10.38,122.80,1001.0,0.11840,0.27760,0.3001,0.14710,...,17.33,184.60,2019.0,0.1622,0.6656,0.7119,0.2654,0.4601,0.11890,NaN
1,842517,M,20.57,17.77,132.90,1326.0,0.08474,0.07864,0.0869,0.07017,...,23.41,158.80,1956.0,0.1238,0.1866,0.2416,0.1860,0.2750,0.08902,NaN
2,84300903,M,19.69,21.25,130.00,1203.0,0.10960,0.15990,0.1974,0.12790,...,25.53,152.50,1709.0,0.1444,0.4245,0.4504,0.2430,0.3613,0.08758,NaN
3,84348301,M,11.42,20.38,77.58,386.1,0.14250,0.28390,0.2414,0.10520,...,26.50,98.87,567.7,0.2098,0.8663,0.6869,0.2575,0.6638,0.17300,NaN
4,84358402,M,20.29,14.34,135.10,1297.0,0.10030,0.13280,0.1980,0.10430,...,16.67,152.20,1575.0,0.1374,0.2050,0.4000,0.1625,0.2364,0.07678,NaN


Little About the Neural Architecture:

input = 30, neuron = 1, ACT = Sigmoid, Output = y_pred

In [583]:
df.shape

(569, 33)

### Dropping Unnecessary Cols (reducing Dimensionality)

In [584]:
df.drop(columns=['id', 'Unnamed: 32'], inplace= True)

In [585]:
df.head()

,diagnosis,radius_mean,texture_mean,perimeter_mean,area_mean,smoothness_mean,compactness_mean,concavity_mean,concave points_mean,symmetry_mean,...,radius_worst,texture_worst,perimeter_worst,area_worst,smoothness_worst,compactness_worst,concavity_worst,concave points_worst,symmetry_worst,fractal_dimension_worst
0,M,17.99,10.38,122.80,1001.0,0.11840,0.27760,0.3001,0.14710,0.2419,...,25.38,17.33,184.60,2019.0,0.1622,0.6656,0.7119,0.2654,0.4601,0.11890
1,M,20.57,17.77,132.90,1326.0,0.08474,0.07864,0.0869,0.07017,0.1812,...,24.99,23.41,158.80,1956.0,0.1238,0.1866,0.2416,0.1860,0.2750,0.08902
2,M,19.69,21.25,130.00,1203.0,0.10960,0.15990,0.1974,0.12790,0.2069,...,23.57,25.53,152.50,1709.0,0.1444,0.4245,0.4504,0.2430,0.3613,0.08758
3,M,11.42,20.38,77.58,386.1,0.14250,0.28390,0.2414,0.10520,0.2597,...,14.91,26.50,98.87,567.7,0.2098,0.8663,0.6869,0.2575,0.6638,0.17300
4,M,20.29,14.34,135.10,1297.0,0.10030,0.13280,0.1980,0.10430,0.1809,...,22.54,16.67,152.20,1575.0,0.1374,0.2050,0.4000,0.1625,0.2364,0.07678


## Train Test Split

In [586]:
# X = df.drop(columns=['diagnosis'])
# y = df['diagnosis']
# X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
X_train, X_test, y_train, y_test = train_test_split(df.iloc[:, 1:], df.iloc[:, 0], test_size=0.2, random_state = 42)

## Preprocessing

### Scaling (neural models perform better when the data are scaled)

In [587]:
scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

In [588]:
X_train

array([[-1.44075296, -0.43531947, -1.36208497, ...,  0.9320124 ,
         2.09724217,  1.88645014],
       [ 1.97409619,  1.73302577,  2.09167167, ...,  2.6989469 ,
         1.89116053,  2.49783848],
       [-1.39998202, -1.24962228, -1.34520926, ..., -0.97023893,
         0.59760192,  0.0578942 ],
       ...,
       [ 0.04880192, -0.55500086, -0.06512547, ..., -1.23903365,
        -0.70863864, -1.27145475],
       [-0.03896885,  0.10207345, -0.03137406, ...,  1.05001236,
         0.43432185,  1.21336207],
       [-0.54860557,  0.31327591, -0.60350155, ..., -0.61102866,
        -0.3345212 , -0.84628745]])

### Label Encoding (Encoding the M/B target dataset)

In [589]:
encoder = LabelEncoder()
y_train = encoder.fit_transform(y_train)
y_test = encoder.transform(y_test)

In [590]:
y_train

array([0, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 1, 1, 0, 0, 1, 1, 1, 0, 0, 0, 1,
       0, 0, 0, 1, 0, 1, 0, 0, 1, 0, 1, 1, 1, 0, 1, 0, 0, 0, 0, 1, 1, 0,
       0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0,
       0, 0, 0, 1, 1, 1, 0, 0, 1, 0, 1, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0,
       1, 0, 1, 1, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 1, 1,
       0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0, 1, 0, 0, 1, 0, 1, 0,
       1, 0, 1, 0, 0, 1, 0, 0, 0, 1, 0, 1, 0, 1, 0, 1, 0, 0, 1, 0, 0, 0,
       0, 1, 0, 0, 0, 1, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0,
       0, 1, 0, 1, 0, 0, 0, 1, 0, 1, 1, 0, 0, 1, 0, 1, 1, 1, 0, 0, 0, 1,
       0, 0, 1, 0, 1, 0, 0, 0, 1, 0, 1, 0, 0, 1, 1, 0, 0, 1, 0, 1, 1, 0,
       1, 1, 0, 0, 1, 1, 1, 0, 0, 0, 0, 1, 0, 1, 1, 1, 1, 0, 0, 0, 0, 0,
       0, 0, 0, 1, 1, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 1, 1, 0, 1, 0, 1,
       0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 1,
       1, 1, 0, 1, 0, 0, 1, 1, 1, 0, 0, 0, 0, 0, 0,

### Numpy arrays to PyTorch tensors (To use PyTorch Training Pipelines)

In [591]:
type(X_train)

numpy.ndarray

So that's a numpy array, all the variables here are numpy arrays. But to use PyTorch Training Pipeline they need to be transferred into PyTorch Tensors.

####**dtype**:
Here, the dataset eventually needs to be float type. This transformation will be done in class `MySimpleNN()`; but the practice is not good, it should be gone here, The commented lines. It intentionally has been done later so one can see both of the ways. And ran into an error before by putting float32 here and float64 in the class `MySimpleNN`. Be careful! And remember float32 takes less memory than float64.

Changed them for nn Module inehritance version.

In [592]:
# transformed_tensor_var = torch.tensor(numpy_array_name)
# transformed_tensor_var = torch.tensor(numpy_arr_name, dtype=torch.float32) # u r gonna use this one

# X_train_tensor = torch.from_numpy(X_train)
# X_test_tensor = torch.from_numpy(X_test)
# y_train_tensor = torch.from_numpy(y_train)
# y_test_tensor = torch.from_numpy(y_test)

# Changing the dtypes
# X_train_tensor = torch.tensor(X_train_tensor, dtype=torch.float64)
# X_test_tensor = torch.tensor(X_test_tensor, dtype=torch.float64)
# y_train_tensor = torch.tensor(y_train_tensor, dtype=torch.float64)
# y_test_tensor = torch.tensor(y_test_tensor, dtype=torch.float64)

X_train_tensor = torch.tensor(X_train, dtype=torch.float64)
X_test_tensor = torch.tensor(X_test, dtype=torch.float64)
y_train_tensor = torch.tensor(y_train, dtype=torch.float64)
y_test_tensor = torch.tensor(y_test, dtype=torch.float64)


In [593]:
type(X_train_tensor)

torch.Tensor

In [594]:
X_train_tensor.shape # (row, col)

torch.Size([455, 30])

In [595]:
y_train_tensor.shape # (row) as col 1 only

torch.Size([455])

## Defining the Model

In [596]:
X_train_tensor.shape

torch.Size([455, 30])

In [597]:
class MySimpleNN():

  def __init__(self, X): # constructor (self(to call w & b), input)
    # initializing weight (size: (30*1)) (to match the matmul operation without Transpose matrix)
    self.weights = torch.rand(X.shape[1], # num of rows = no of input/cols (num of col in input dataset)
                              1, # num of col = 1 # as all features going to only one neuron
                              dtype=torch.float64, # must be float, gradient of int type can't be calculated
                              requires_grad=True)  # keeps track

    # initializing bias (1*1)
    self.bias = torch.zeros(1, # 1 element matrix for the only one neuron
                            dtype=torch.float64,
                            requires_grad=True)

  def forward(self, X): # (self, input); Z = WX + b
    z = torch.matmul(X, self.weights) + self.bias
    y_pred = torch.sigmoid(z)
    return y_pred

  def loss_function(self, y_pred, y):

    # Clamp predictions to avoid log(0) or log(1) respective to undefined or 0.
    epsilon = 1e-7
    y_pred = torch.clamp(y_pred, epsilon, 1 - epsilon) # var = torch.clamp(var, min, max)

    # Calculate loss
    loss = -(y_train_tensor * torch.log(y_pred) + (1 - y_train_tensor) * torch.log(1 - y_pred)).mean()
    return loss


You can check the weights and biases after training the pipeline

### Important Parameters (lr and epoch)

In [598]:
learning_rate = 0.1
epochs = 25

## Training Pipeline

In [599]:
# create model
model = MySimpleNN(X_train_tensor) # training the model on train dataset

# define loop
for epoch in range(epochs):

  # forward pass
  y_pred = model.forward(X_train_tensor)

  # loss calculate
  loss = model.loss_function(y_pred, y_train_tensor)

  # backward pass
  loss.backward()

  # parameters update: so it renew every time ~ = the loop doesnt change the values each time of w_old.
  with torch.no_grad():
    model.weights -= learning_rate * model.weights.grad
    model.bias -= learning_rate * model.bias.grad

  # zero gradients
  model.weights.grad.zero_()
  model.bias.grad.zero_()

  # print loss in each epoch
  print(f'Epoch: {epoch + 1}, Loss: {loss.item()}') # epoch + 1 as count starts from zero

Epoch: 1, Loss: 3.921500754352041
Epoch: 2, Loss: 3.801013960056209
Epoch: 3, Loss: 3.678026366202628
Epoch: 4, Loss: 3.5517447650142078
Epoch: 5, Loss: 3.4195645689958227
Epoch: 6, Loss: 3.2813393868788667
Epoch: 7, Loss: 3.136037068034449
Epoch: 8, Loss: 2.9888102599785826
Epoch: 9, Loss: 2.838044771817683
Epoch: 10, Loss: 2.679854239027117
Epoch: 11, Loss: 2.518795505616153
Epoch: 12, Loss: 2.3529966224811303
Epoch: 13, Loss: 2.186373585470229
Epoch: 14, Loss: 2.0215553510988693
Epoch: 15, Loss: 1.8597094282404356
Epoch: 16, Loss: 1.7025470905702917
Epoch: 17, Loss: 1.552872813580979
Epoch: 18, Loss: 1.4077817472551328
Epoch: 19, Loss: 1.2767867136020667
Epoch: 20, Loss: 1.1617881087370574
Epoch: 21, Loss: 1.0642398846375147
Epoch: 22, Loss: 0.9847337424495236
Epoch: 23, Loss: 0.9226172202634555
Epoch: 24, Loss: 0.875887795092224
Epoch: 25, Loss: 0.8415748713354926


In [600]:
model.weights
model.bias

tensor([-0.1354], dtype=torch.float64, requires_grad=True)

## Evaluation

In [601]:
y_pred # only printing the 30th element to avoid the whole list.

tensor([[0.2787],
        [0.5880],
        [0.4149],
        [0.7741],
        [0.3664],
        [0.7548],
        [0.8469],
        [0.2560],
        [0.2493],
        [0.3376],
        [0.5170],
        [0.4272],
        [0.5851],
        [0.4054],
        [0.1656],
        [0.5675],
        [0.5428],
        [0.5516],
        [0.1214],
        [0.4355],
        [0.0693],
        [0.3585],
        [0.4357],
        [0.8979],
        [0.3842],
        [0.3122],
        [0.4665],
        [0.4541],
        [0.1749],
        [0.0647],
        [0.3500],
        [0.3550],
        [0.4895],
        [0.7871],
        [0.1861],
        [0.1827],
        [0.3049],
        [0.2105],
        [0.4261],
        [0.5617],
        [0.3757],
        [0.9421],
        [0.8577],
        [0.4163],
        [0.2479],
        [0.1515],
        [0.3684],
        [0.1993],
        [0.6970],
        [0.5814],
        [0.7688],
        [0.3793],
        [0.2947],
        [0.4055],
        [0.8241],
        [0

Notice how y_pred is also tracking. We've already seen this in the computational graph. After sigmoid operation we get y_pred. So y_pred is also in requires_grad.

In [602]:
# model evaluation
with torch.no_grad(): # y_pred is also tracking so it will change the value with each loop; that's why the whole thing needs to be under no_grad()
  y_pred = model.forward(X_test_tensor) # getting the y_pred tensor
  y_pred = (y_pred > 0.9).float() # y_pred True if greater than 0.9. Then applying float to convert it to 0/1 (as accuracy range is 0 to 1)
  # accuracy = all true assumptions/ all assumptions
  accuracy = (y_pred == y_test_tensor).float().mean() # accuracy = if y_pred = y (T/F); then converting to float. Then the mean of all the 1 + 0! (! Dividing by all assumptions means mean)
  print(f'Accuracy: {accuracy.item()}')


Accuracy: 0.6120344996452332


In [603]:
accuracy

tensor(0.6120)

Only writing accuracy will give u a result like this. To access the values from the tensor, we use item. Like
`
accuracy.item()
`

Accuracy:
- 0.6207
- 0.6120 (after adding those 4 transformed `dtype` lines which are commented currently)
- 0.6185 (After running multiple times replicating the first situation, cant understand why)
- 0.6163 (After re running again)

In [604]:
y_pred # y_pred = (y_pred > 0.9 ).float()

tensor([[0.],
        [0.],
        [0.],
        [0.],
        [0.],
        [0.],
        [0.],
        [0.],
        [0.],
        [0.],
        [0.],
        [0.],
        [0.],
        [0.],
        [0.],
        [1.],
        [0.],
        [0.],
        [0.],
        [0.],
        [0.],
        [0.],
        [0.],
        [0.],
        [0.],
        [0.],
        [0.],
        [0.],
        [0.],
        [0.],
        [0.],
        [0.],
        [0.],
        [0.],
        [0.],
        [0.],
        [0.],
        [0.],
        [0.],
        [0.],
        [0.],
        [0.],
        [0.],
        [0.],
        [0.],
        [0.],
        [0.],
        [0.],
        [0.],
        [0.],
        [0.],
        [0.],
        [0.],
        [0.],
        [0.],
        [0.],
        [0.],
        [0.],
        [0.],
        [0.],
        [0.],
        [0.],
        [0.],
        [0.],
        [0.],
        [0.],
        [0.],
        [1.],
        [0.],
        [0.],
        [1.],
      

# 12.7 By inheriting nn module (Version 2)

We don't need to change anything from here as we are not changing the neural network, only making the code easier by integrating nn module (a feature in PyTorch).

### Importing nn Module

In [605]:
import torch.nn as nn

## Defining the model (nn module)

In [606]:
class MySimpleNN2 (nn.Module): # inheriting nn module in this class
  def __init__(self, num_features): # constructor # num_features = input_tensor[1] == num of inputs
    super().__init__() # nn module attribute involvement

    self.linear = nn.Linear (num_features, 1) # nn.Linear Transformation Operation (input nodes, output nodes) # 30 inputs and 1 direct output
    self.sigmoid = nn.Sigmoid () # the ACT was Sigmoid, so bringing that

  def forward (self, features):
    output = self.linear(features) # Linear Transformation Output
    # print(f"Linear Transformation Output: {output}")
    output = self.sigmoid(output) # it will save memory as being saved on the same storage
    # print(f"\nActivation Function Ouput: {output}")

  # do not need to write loss function def here, as nn has it's own BCELoss Function

    return output



Declaring learning rate and epochs for later.

In [607]:
learning_rate = 0.1
epochs = 25

### Loss Function

In [608]:
loss_function = nn.BCELoss()

In [609]:
type(loss_function)

torch.nn.modules.loss.BCELoss

The `type(loss_function)` gives equation of the Loss Function.

## Training Pipeline

`model = model.double()`

double = float64, float = float32.

- class 1: MySimpleNN -> float64 inside code; to match that it was float64 in numpy array to tensor transformation.

- class 2: MySimpleNN2 -> Default NN Module has float32, so it ran into an error. That's why I had to add an extra line here, also I could change the code and make it float32 (Float) instead of float64 (Double).

- summary: it presents the report in float32 default.

####**Use float32 for the better**

In [610]:
# creating the model
model = MySimpleNN2(X_train_tensor.shape[1]) # model = class_name (input) # input = col of X_train_tensor
model = model.double() # Cast model parameters to float64 to match input tensor

# optimizing
optimizer = torch.optim.SGD(model.parameters(), # SGD > GD (Sochastic Gradient Descent)
                            lr=learning_rate)  # lr = learning rate

# Defining the Loop
for epoch in range(epochs):

  # forward pass
  y_pred = model(X_train_tensor)

  # loss calculation
  loss = loss_function(y_pred, # y_pred.shape : 2D -> (batch_size, 1)
                       y_train_tensor.reshape(-1, 1)) # y_train : 1D -> (batch_size, )
                       # here reshape(-1,1): 1 is the no of col, -1 is saying to caldulate the necessary rows to store y_train in 1 col.
                       # As 2D can't be calculated with 1D, we are reshaping the y_train_tensor here.
  # ouput of loss: tensor(0.1748, dtype=torch.float64, grad_fn=<BinaryCrossEntropyBackward0>)

  # zero gradient : making everything zero before started backward passing , so yeah before backward pass
  optimizer.zero_grad()

  # backward : computes dL/dW and attaches them to weight.grad
  loss.backward() # why loss.backward? -> think about the computation graph (forward -> prediction -> loss -> backward)

  # updating parameters : wnre = wold - lr * dL/dw
  optimizer.step()

  # printing loss in each epoch
  print(f'Epoch: {epoch + 1}, Loss: {loss.item()}')


Epoch: 1, Loss: 0.5311890736311581
Epoch: 2, Loss: 0.4457340007675406
Epoch: 3, Loss: 0.39066022424130264
Epoch: 4, Loss: 0.35202163947535065
Epoch: 5, Loss: 0.3232356922378067
Epoch: 6, Loss: 0.3008278897187299
Epoch: 7, Loss: 0.2827962941725559
Epoch: 8, Loss: 0.26790642895040895
Epoch: 9, Loss: 0.2553551119556054
Epoch: 10, Loss: 0.24459659526446997
Epoch: 11, Loss: 0.23524652771750873
Epoch: 12, Loss: 0.2270259582908228
Epoch: 13, Loss: 0.2197271789846311
Epoch: 14, Loss: 0.21319207765759451
Epoch: 15, Loss: 0.20729796155715358
Epoch: 16, Loss: 0.20194800510362243
Epoch: 17, Loss: 0.19706465072972187
Epoch: 18, Loss: 0.19258494792338526
Epoch: 19, Loss: 0.18845719562501387
Epoch: 20, Loss: 0.18463848025048793
Epoch: 21, Loss: 0.18109284124986108
Epoch: 22, Loss: 0.17778988416460637
Epoch: 23, Loss: 0.17470371794885695
Epoch: 24, Loss: 0.1718121307251743
Epoch: 25, Loss: 0.1690959432418986


### Parameters Checking

In [616]:
model.parameters()

<generator object Module.parameters at 0x7b65125617e0>

## Evaluation

In [611]:
with torch.no_grad():
  y_pred = model.forward(X_test_tensor)
  y_pred= (y_pred > 0.9).float()
  accuracy = (y_pred == y_test_tensor).float().mean()
  print(f" Accuracy = {accuracy.item()}")

 Accuracy = 0.5560172200202942


## Summary: torchinfo

- install dependencies
- calling summary

In [612]:
!pip install torchinfo

Importing the dependency

In [613]:
from torchinfo import summary

In [615]:
summary(model, input_size=(455, 30), dtypes=[torch.float64]) # usually summary creates default float32, so ran into an error again.

Layer (type:depth-idx)                   Output Shape              Param #
MySimpleNN2                              [455, 1]                  --
├─Linear: 1-1                            [455, 1]                  31
├─Sigmoid: 1-2                           [455, 1]                  --
Total params: 31
Trainable params: 31
Non-trainable params: 0
Total mult-adds (Units.MEGABYTES): 0.01
Input size (MB): 0.11
Forward/backward pass size (MB): 0.01
Params size (MB): 0.00
Estimated Total Size (MB): 0.12